datalad/support/gitrepo.py

```diff
@@ -3007,16 +2989,18 @@ def _get_cache_key(label, paths, ref, untracked=None):
             _cache[key] = from_state
 
         status = OrderedDict()
-        for f, to_state_r in to_state.items():
+        # for all paths we know now or knew before
+        for f in set(from_state).union(to_state):
             props = self._diffstatus_get_state_props(
                 f,
-                from_state.get(f, None),
-                to_state_r,
+                from_state.get(f),
+                to_state.get(f),
                 # are we comparing against a recorded commit or the worktree
                 to is not None,
-                # if we have worktree modification info, report if
-                # path is reported as modified in it
-                modified and f in modified,
+                # if we have worktree modification info, pass type(change)
+                # report on, if there is any
+                False if modified is None
```

Comment:
It is right there, on line 2978 in the `else` clause. It indicates that there cannot be modifications in the worktree, because the worktree is not part of the comparison.